### PR TITLE
Causal response function for Ricardo graph

### LINKED ISSUE #134: Function for evaluating d
This will be similar to the $r$ -builder in #129; we'll need a function `d_builder` (or `d_maker`) that takes in a graph $G$ representing Ricardo's graph. Using this information, it needs to evaluate the function $d(x, l; \theta)$ as defined in Ricardo's document. We're hard-coding the functional form of this method, for now.

The function $d(x, l; \theta)$ is defined as

$$
d(x, l; \theta) = \mathbb{E}[Y \vert \text{do}(X=x), L=l]
= \int f_Y(u_y, x, l; \theta_Y)p_N(u_y) \mathrm{d} u_y
$$

where $p_N$ is the PDF of a normal distribution with mean 0 and standard deviation 1.

We need to write a function

```python
def d_builder(graph, n_points) -> Callable[[dict, dict], float]:
```

that takes Ricardo's graph `graph` as an input. Given the form of the integral it makes the most sense for us to start by hard-coding in use of the `causalprog.quadrature.UniformWeightMonteCarloGaussianQuadrature` method, and hence the `n_points` input to `d_builder` should simply be used to create an instance of this quadrature class, for evaluating the integral. Using `UniformWeightMonteCarloGaussianQuadrature.integrate` with an integrand of $f_Y$ should then be sufficient to evaluate $d$ (see the class docstring for more information).

`d_builder` needs to return a function `d` with signature `d(u_yxl, theta)`. `u_yxl` should be a dictionary with keys `"u_y"`, `"x"`, `"l"` that correspond to the argument values $u_y$, $x$, $l$ that $d$ takes. `theta` should be a dictionary with keys such as `"theta_y"`, `theta_x`, etc, and corresponds to the $\theta$ argument to $d$.

### PR DESCRIPTION
**Pull request by samjmolyneux**
Closes #134 

* Separated `vectorise_over_dict_args` out so it can be reused in the tests.
* Added `build_causal_response_function(...)` to `src/causalprog/graph/ricardo.py`.
* Added three tests to `tests/test_integration/test_ricardo_causal_response_function.py`:

  1. `test_fy_independent_of_uy(...)` uses the fact that if $`f_Y`$ is independent of $`u_y`$, then

     $`d(x,l;\theta)=\int_{-\infty}^{\infty}f_Y(u_y,x,l;\theta_Y)\,p_{\mathcal{N}}(u_y;0,1)\,\mathrm{d}u_y`$<br>
     $`=f_Y(x,l;\theta_Y)\int_{-\infty}^{\infty}p_{\mathcal{N}}(u_y;0,1)\,\mathrm{d}u_y`$<br>
     $`=f_Y(x,l;\theta_Y).`$

  2. `test_causal_response_matches_standard_normal_moments(...)` uses a quadratic response function of the form

     $`f_Y(u_y,x,l;\theta_Y)=\theta_{\mathrm{quadratic}}u_y^2+\theta_{\mathrm{linear}}u_y+\theta_xx+\theta_l\sum_i l_i.`$

     Since $`U_Y\sim\mathcal{N}(0,1)`$, its first two moments are

     $`\mathbb{E}[U_Y]=0`$ and $`\mathbb{E}[U_Y^2]=1.`$

     Therefore,

     $`d(x,l;\theta)=\int_{-\infty}^{\infty}f_Y(u_y,x,l;\theta_Y)\,p_{\mathcal{N}}(u_y;0,1)\,\mathrm{d}u_y`$<br>
     $`=\theta_{\mathrm{quadratic}}\mathbb{E}[U_Y^2]+\theta_{\mathrm{linear}}\mathbb{E}[U_Y]+\theta_xx+\theta_l\sum_i l_i`$<br>
     $`=\theta_{\mathrm{quadratic}}+\theta_xx+\theta_l\sum_i l_i.`$
  3. `test_causal_response_is_jittable_and_differentiable` does as it says. 
  
For `test_causal_response_matches_standard_normal_moments(...)`, I used `n_points=1,000,000`, adjusted `rtol`  because the errors were too large and added batching to overcome memory issues. 
It works relatively okay at 1,000,000 but after further investigation that just seems to be a local minimum. I have included a graph of `n_points` vs largest absolute error below. 

Also, I was getting inf errors until I started using 64 bit jax. 
  
output (1)

### CODE REVIEW

File: src/causalprog/graph/ricardo.py [hunk 1]
```diff
@@ -187,3 +187,101 @@ def _r(
...
+    =
+    \int_{-\infty}^{\infty}
+    f_Y(u_y, x, l; \theta_Y)
+    p_N(u_y)
+    \,\mathrm{d}u_y.
+    $$
+
+    The returned callable has signature ``d(xl, model_params)``, where
+    ``xl`` contains the fixed values of ``x`` and ``l``. The latent variable
+    ``u_y`` is supplied internally by the quadrature rule.
```
Comment (willGraham01) [inline src/causalprog/graph/ricardo.py]:
```suggestion
    The returned callable has signature `d(xl, model_params)`, where
    `xl` contains the fixed values of `x` and `l`. The latent variable
    `u_y` is supplied internally by the quadrature rule.
```

Minor nitpick but since we're using mkdocs and not sphinx, inline code only needs a single backtick rather than double backticks.

File: src/causalprog/graph/ricardo.py [hunk 2]
```diff
@@ -187,3 +187,101 @@ def _r(
...
+    ) -> float | NDArray:
+        """Evaluate the outcome function at one quadrature point."""
+        return f_y(
+            {
+                "u_y": u_y,
+                "x": xl["x"],
+                "l": xl["l"],
+            },
+            model_params["theta_y"],
+        )
```
Comment (willGraham01) [inline src/causalprog/graph/ricardo.py]:
```suggestion
    def _integrand(
        u_y: float,
        xl: dict[str, float | NDArray],
        model_params: dict[str, ModelParam],
    ) -> float | NDArray:
        """Evaluate the outcome function at one quadrature point."""
        return node_y.compute(
            {
                "u_y": u_y,
                "x": xl["x"],
                "l": xl["l"],
            },
            model_params["theta_y"],
        )
```

`f_y` is kind of redundant here (though I imagine it likely helped to have it when first writing the method). We need `_integrand` of course so that we can pass into the quadrature rule, but I think we can save ourselves a few lines otherwise.

File: src/causalprog/graph/ricardo.py [hunk 3]
```diff
@@ -187,3 +187,101 @@ def _r(
         )
 
     return _r
+
+
+def build_causal_response_function(
+    graph: Graph,
+    quadrature: QuadratureMethod,
+) -> Callable:
```
Comment (willGraham01) [inline src/causalprog/graph/ricardo.py]:
```suggestion
) -> MLPAlias:
```

`MLPAlias` is kind of misleading and I should probably update it's name, but it will work as the return typehint here.

Now that I mention it, the same change should be made for `build_regression_function`'s typehint too, so feel free to do that while you're at it 😅

File: tests/test_integration/_helpers.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+from collections.abc import Iterable
```
Comment (willGraham01) [inline tests/test_integration/_helpers.py]: To save on adding to the `_helpers.py` file (keeping in mind #158), I recommend moving this function into `tests/fixtures/jax.py`. You can then turn it into a fixture via:

```python
@pytest.fixture
def vectorise_over_dict_args() -> Callable[Concatenate[MLPAlias, ...], MLPAlias]:
    """Vectorise a pure function of dictionary arguments across the dictionary keys.

    This is essentially a wrapper around iterative applications of `jax.vmap` with the
    appropriate `in_axes` specified. The net effect is that if the input `f` was
    being called with a dictionary argument, whose keys were scalar-valued, the returned
    function can be called with the same dictionary argument whose keys are
    vector-valued, and returns a vector-valued output.

    Note that all vmap-ing is done along axis 0. If you want to pass in vector-values
    for some of the dictionary key inputs, ensure that they are aligned along the
    correct axis (each _row_ should be one value of the input).
    """
    def _inner(f: MLPAlias, *dict_keys: Iterable[str]) -> MLPAlias:
        vec_f = f
        all_keys = [key for key_list in dict_keys for key in key_list]
        for key in all_keys:
            vec_f = jax.vmap(
                vec_f,
                in_axes=tuple(
                    {k: None if k != key else 0 for k in arg_keys} for arg_keys in dict_keys
                ),
            )
        return vec_f

    return _inner
```

If you want to use it in a test, just add it to the test arguments like you would any other fixture.

(`Concatenate` is importable from `typing`)

File: tests/test_integration/test_ricardo/test_causal_response_function.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+    For U_Y ~ N(0, 1),
+
+    E[U_Y] = 0
+    and
+    E[U_Y**2] = 1.
+
+    The Cartesian product of input values is evaluated in batches to limit
+    peak memory usage.
+    """
```
Comment (willGraham01) [inline tests/test_integration/test_ricardo/test_causal_response_function.py:131]:
```suggestion
    r"""Compare d against known moments of a standard-normal latent variable.

    For $U_Y \sim \mathcal{N}(0, 1)$,
    
    $$
    \mathcal{E}[U_Y] = 0,
    \quad\text{ and }\quad
    \mathcal{E}{U_Y^2} = 1.
    $$
    
    The Cartesian product of input values is evaluated in batches to limit
    peak memory usage.
    """
```

$\LaTeX$-ifying

File: tests/test_integration/test_ricardo_causal_response_function.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    ) -> jax.Array:
+        return d(xl, {"theta_y": theta_y})
+
+    def d_analytic(
+        xl: dict,
+        theta_y: dict,
+    ) -> jax.Array:
+        return theta_y["x_scale"] * jnp.exp(-(xl["x"] ** 2)) + theta_y[
+            "l_scale"
+        ] * jnp.sum(xl["l"])
```
Comment (willGraham01) [inline tests/test_integration/test_ricardo_causal_response_function.py]:
```suggestion
    d_analytic = f_y
```

You could even drop defining `d_analytic` everywhere and just use `f_y` in it's place. Even though `f_y` takes a variable called `u_yxl` as it's first argument, Python doesn't actually care if we only pass in `xl` since we're assuming independence from `u_y` in this test.

File: tests/test_integration/test_ricardo_causal_response_function.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    analytic_grid = vectorise_over_dict_args(
+        d_analytic,
+        xl.keys(),
+        theta_y.keys(),
+    )
+
+    actual = d_grid(xl, theta_y)
+    expected = analytic_grid(xl, theta_y)
+
+    assert jnp.allclose(actual, expected)
```
Comment (willGraham01) [inline tests/test_integration/test_ricardo_causal_response_function.py]:
```suggestion
    d_from_theta_y = vectorise_over_dict_args(
        d_from_theta_y,
        xl.keys(),
        theta_y.keys(),
    )
    d_analytic = vectorise_over_dict_args(
        d_analytic,
        xl.keys(),
        theta_y.keys(),
    )

    assert jnp.allclose(d_from_theta_y(xl, theta_y), d_analytic(xl, theta_y))
```

Just a style comment to remove a few lines and extra variable names.

File: tests/test_integration/test_ricardo_regression_function.py [hunk 8]
```diff
@@ -67,8 +42,9 @@ def test_fy_independent_of_uy(
     f_m: MLPAlias = lambda czl, theta_r: czl["c"] * theta_r,
     f_ux: MLPAlias = lambda xzl, theta_x: xzl["x"] * theta_x[0],
     f_y: MLPAlias = lambda xu_y, theta_y: theta_y[0] * jnp.exp(-(xu_y["x"] ** 2)),
-    r_analytic: MLPAlias = lambda xzl, theta: theta["theta_y"][0]
-    * jnp.exp(-(xzl["x"] ** 2)),
+    r_analytic: MLPAlias = lambda xzl, theta: (
+        theta["theta_y"][0] * jnp.exp(-(xzl["x"] ** 2))
+    ),
 ) -> None:
```
Comment (willGraham01) [inline tests/test_integration/test_ricardo_regression_function.py:16]: Just a note here that if you make `vectorise_over_dict_args` a fixture, you'll need to include it in the argument list here too

File: tests/test_integration/test_ricardo/test_causal_response_function.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+def test_causal_response_matches_standard_normal_moments(
+    rng_key: jax.Array,
+    jax_enable_x64,  # noqa: ARG001
+    n_points: int = 1_000_000,
+    n_eval_pts_per_dim: int = 4,
+    eval_range: tuple[float, float] = (-10.0, 10.0),
+    grid_batch_size: int = 64,
+    atol: float = 0.0,
+    rtol: float = 0.01,
```
Comment (willGraham01) [inline tests/test_integration/test_ricardo/test_causal_response_function.py:118]: The `rtol` seems quite big given the number of sample points `n_points`. 1 million sample points should be more than enough to perform the integral to suitable accuracy. Though from your graph it looks like this is genuinely the performance our MC integrator are giving us... though I wonder if this is a floating-point-arithmetic thing (if you were getting `inf` errors beforehand, it feels like it could be some kind of under/overflow).

We should look into this if we find out things can't be made tighter. Is this just a case that there's some small region of the grid where things are terrible (and in other places a tighter `rtol` would be fine) or a case that it's genuinely bad everywhere?

**Review by willGraham01 — CHANGES_REQUESTED** [review]
Just some tidying for the test suite but otherwise great job (especially on finding two actually meaningful cases to test that also aren't too complex)! There is one test though where the relative errors seem quite loose.

Also, don't forget to change the PR title (it's currently using your branch name).

File: tests/test_integration/test_ricardo/test_causal_response_function.py (hunk 9 above)
Comment (willGraham01) [reply tests/test_integration/test_ricardo/test_causal_response_function.py:118]: Now referenced in #170

**Review by willGraham01 — APPROVED** [review]
